File: sdk/python/gateorix/adapter.py
"""
Gateorix runtime adapter for Python.

Implements the adapter protocol over stdio (newline-delimited JSON).
The host core spawns this process and communicates by writing JSON
messages to stdin and reading responses from stdout.
"""

import json
import sys
from typing import Any, Callable

CommandHandler = Callable[[dict[str, Any]], Any]


class GateorixAdapter:
    """
    Python runtime adapter for Gateorix.

    Register command handlers and call `run()` to start the stdio
    message loop.

    Example:
        adapter = GateorixAdapter()

        @adapter.command("greet")
        def greet(payload):
            return {"message": f"Hello, {payload.get('name', 'World')}!"}

        adapter.run()
    """

    def __init__(self) -> None:
        self._handlers: dict[str, CommandHandler] = {}

    def command(self, name: str) -> Callable[[CommandHandler], CommandHandler]:
        """Decorator to register a command handler."""
        def decorator(func: CommandHandler) -> CommandHandler:
            self._handlers[name] = func
            return func
        return decorator

    def run(self) -> None:
        """Start the stdio message loop. Blocks until stdin is closed."""
        for line in sys.stdin:
            line = line.strip()
            if not line:
                continue

            try:
                request = json.loads(line)
            except json.JSONDecodeError:
                self._send_error("unknown", "invalid JSON")
                continue

            req_id = request.get("id", "unknown")
            channel = request.get("channel", "")
            payload = request.get("payload", {})

            # Extract action from channel (e.g. "runtime.greet" → "greet")
            action = channel.split(".")[-1] if "." in channel else channel

            handler = self._handlers.get(action)
            if handler is None:
                self._send_error(req_id, f"unknown command: {action}")
                continue

            try:
                result = handler(payload)
                self._send_ok(req_id, result)
            except Exception as exc:
                self._send_error(req_id, str(exc))

    def _send_ok(self, req_id: str, payload: Any) -> None:
        response = {"id": req_id, "ok": True, "payload": payload}
        sys.stdout.write(json.dumps(response) + "\n")
        sys.stdout.flush()

    def _send_error(self, req_id: str, message: str) -> None:
        response = {"id": req_id, "ok": False, "payload": {"error": message}}
        sys.stdout.write(json.dumps(response) + "\n")
        sys.stdout.flush()

File: sdk/go/gateorix/adapter.go
// Package gateorix provides a runtime adapter SDK for building Gateorix
// desktop app backends in Go. It implements the adapter protocol over
// stdio (newline-delimited JSON).
//
// Example:
//
//	package main
//
//	import "github.com/gateorix/gateorix/sdk/go/gateorix"
//
//	func main() {
//	    adapter := gateorix.NewAdapter()
//	    adapter.Command("greet", func(payload map[string]any) (any, error) {
//	        name, _ := payload["name"].(string)
//	        if name == "" { name = "World" }
//	        return map[string]string{"message": "Hello, " + name + "!"}, nil
//	    })
//	    adapter.Run()
//	}
package gateorix

import (
	"bufio"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
)

// CommandHandler is a function that handles a Gateorix adapter command.
// It receives the payload and returns a result or an error.
type CommandHandler func(payload map[string]any) (any, error)

// Request represents a message from the host core to the sidecar.
type Request struct {
	ID      string         `json:"id"`
	Channel string         `json:"channel"`
	Payload map[string]any `json:"payload"`
}

// Response represents a message from the sidecar back to the host core.
type Response struct {
	ID      string `json:"id"`
	OK      bool   `json:"ok"`
	Payload any    `json:"payload"`
}

// Adapter is the main entry point for a Gateorix Go backend.
type Adapter struct {
	handlers map[string]CommandHandler
}

// NewAdapter creates a new adapter instance.
func NewAdapter() *Adapter {
	return &Adapter{
		handlers: make(map[string]CommandHandler),
	}
}

// Command registers a handler for the given channel name.
// The channel is matched against the "channel" field in incoming requests
// (e.g. "runtime.greet" matches if you register "greet" — the "runtime."
// prefix is stripped automatically).
func (a *Adapter) Command(name string, handler CommandHandler) {
	a.handlers[name] = handler
}

// dispatch routes a request to the appropriate handler.
func (a *Adapter) dispatch(req Request) Response {
	// Strip "runtime." prefix if present
	channel := req.Channel
	if len(channel) > 8 && channel[:8] == "runtime." {
		channel = channel[8:]
	}

	handler, ok := a.handlers[channel]
	if !ok {
		return Response{
			ID: req.ID,
			OK: false,
			Payload: map[string]string{
				"error": fmt.Sprintf("unknown command: %s", req.Channel),
			},
		}
	}

	result, err := handler(req.Payload)
	if err != nil {
		return Response{
			ID: req.ID,
			OK: false,
			Payload: map[string]string{
				"error": err.Error(),
			},
		}
	}

	return Response{
		ID:      req.ID,
		OK:      true,
		Payload: result,
	}
}

// Run starts the stdio message loop, reading JSON requests from stdin
// and writing JSON responses to stdout. Blocks until stdin is closed.
func (a *Adapter) Run() {
	scanner := bufio.NewScanner(os.Stdin)
	// Allow large messages (up to 10 MB)
	scanner.Buffer(make([]byte, 0, 64*1024), 10*1024*1024)

	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}

		var req Request
		if err := json.Unmarshal([]byte(line), &req); err != nil {
			resp := Response{
				ID: "",
				OK: false,
				Payload: map[string]string{
					"error": fmt.Sprintf("invalid JSON: %v", err),
				},
			}
			out, _ := json.Marshal(resp)
			fmt.Fprintln(os.Stdout, string(out))
			continue
		}

		resp := a.dispatch(req)
		out, err := json.Marshal(resp)
		if err != nil {
			resp = Response{
				ID: req.ID,
				OK: false,
				Payload: map[string]string{
					"error": fmt.Sprintf("marshal error: %v", err),
				},
			}
			out, _ = json.Marshal(resp)
		}
		fmt.Fprintln(os.Stdout, string(out))
	}
}

// RunHTTP starts an HTTP server on the given address (e.g. ":3001").
// The host core sends POST requests with JSON bodies to any path.
func (a *Adapter) RunHTTP(addr string) error {
	mux := http.NewServeMux()

	mux.HandleFunc("/invoke", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
			return
		}

		var req Request
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			http.Error(w, "invalid JSON", http.StatusBadRequest)
			return
		}

		resp := a.dispatch(req)
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(resp)
	})

	// Health check
	mux.HandleFunc("/health", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		fmt.Fprintln(w, `{"status":"ok"}`)
	})

	log.Printf("Gateorix Go adapter listening on %s", addr)
	return http.ListenAndServe(addr, mux)
}
